lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -422,6 +422,121 @@ static bool sinkCode(Function *F) {
   return changed;
 }
 
+/// \returns True if node A may depend on the result of B. The relationship
+/// between the nodes does not have to be direct. For example, A can depend on
+/// X which depends on B. In that case the method needs to return True.
+/// Check the use-def dependency up to a depth of \p depth.
+static bool mayDepend(Node *A, Node *B, unsigned depth = 6) {
+  // We define the identify as a dependency.
+  if (A == B) {
+    return true;
+  }
+
+  // A does not depend on anything.
+  if (A->getNumInputs() == 0) {
+    return false;
+  }
+
+  // B has no users. Nothing can depend on it.
+  if (B->getNumResults() == 0) {
+    return false;
+  }
+
+  // We can't continue the search. Assume that the nodes depend on one another.
+  if (depth == 0) {
+    return true;
+  }
+
+  // Check all inputs of A. None of them may depend on B.
+  for (int i = 0, e = A->getNumInputs(); i < e; i++) {
+    auto *input = A->getNthInput(i).getNode();
+    // The inputs of A must not depend on B.
+    if (mayDepend(input, B, depth - 1)) {
+      return true;
+    }
+  }
+
+  // We checked all inputs of A and none of them depend on B.
+  return false;
+}
+
+/// \returns True if the node \p N depends on any of the values in \p list, or
+/// if any of the values in list depend on \p N.
+static bool mayDependOnAny(llvm::ArrayRef<NodeValue> list, Node *N) {
+  for (auto ll : list) {
+    if (mayDepend(ll.getNode(), N) || mayDepend(N, ll.getNode())) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+// Merge several two or more multiple matrix multiplications into a single
+// large matmul. The large matmul is more likely to utilize the hardware. The
+// result of the big matmul is the concatenated results.
+//
+//            ____      _________        _________
+//   ----    |    |    |         |     M|  A * C  |
+// M| A  |  T| B  | * K|    C    | =    |---------|
+//   ---- ,  |    |    |         |     T|  B * C  |
+//    K       ----      ---------        ---------
+//             K            R                R
+static void mergeMatMul(Function *F) {
+  auto &nodes = F->getNodes();
+
+  // These two maps record the list of matrix multipliers that use each node
+  // value either as a right-hand-side user or a left-hand-user.
+  llvm::DenseMap<Node *, std::vector<MatMulNode *>> rightMatrixUsers;
+  llvm::DenseMap<Node *, std::vector<MatMulNode *>> leftMatrixUsers;
+
+  // Collect the list of nodes that are used by the matrix multiplier.
+  for (auto const &node : nodes) {
+    if (auto *MM = dyn_cast<MatMulNode>(node)) {
+      rightMatrixUsers[MM->getRHS().getNode()].push_back(MM);
+      leftMatrixUsers[MM->getLHS().getNode()].push_back(MM);
+    }
+  }
+
+  // Merge RHS matrices.
+  for (auto &it : rightMatrixUsers) {
+    auto &MMs = it.second;
+
+    // Collects the LHS values to merge.
+    std::vector<NodeValue> LHS;
```

Comment:
we only can merge same {S,O} LHS values

Reply:
@rdzhabarov You are right. Good catch. I will fix this.

Reply:
This is a good point, however I do want to point out that this optimization occurs both pre and post quantization/profiling. I think it is most likely to apply to mat muls pre-quantization (on the very first optimization pass), correct? Which brings up an interesting point -- this mat mul merging will force mat muls that individually/previously would have their own quantization parameters to instead now all share the same parameters, meaning they will likely have worse precision at the cost of better utilization. (Of course, this is the tradeoff for quantization in the first place 😄)